File: GoBlackJack/cardpile.go
package main

import (
	"math/bits"
	"time"
)

var state = uint64(time.Now().Unix())

// From https://github.com/lemire/testingRNG
func wyrand() uint64 {
	state += 0xa0761d6478bd642f
	hi, lo := bits.Mul64(state^0xe7037ed1a0b428db, state)
	return uint64(hi ^ lo)
}

// use nearly divisionless technique found here https://github.com/lemire/FastShuffleExperiments
func rand_range(s uint64) uint64 {
	x := wyrand()
	mhi, mlo := bits.Mul64(x, s)
	l := mlo
	if l < s {
		t := -s % s
		for l < t {
			x = wyrand()
			mhi, mlo = bits.Mul64(x, s)
			l = mlo
		}
	}
	return mhi
}

// CardPile class
type CardPile struct {
	MCards         []*Card
	mOriginalCards []*Card
}

// NewCardPile constructor
func NewCardPile(numofdecks int32) CardPile {
	var cp CardPile
	for x := int32(0); x < numofdecks; x++ {
		temp := NewDeck()
		cp.mOriginalCards = append(cp.mOriginalCards, temp.MCards...)
	}

	cp.Refresh()
	return cp
}

// Print the cards
func (c *CardPile) Print() string {
	output := ""
	for _, card := range c.MCards {
		output += card.Print()
	}
	return output
}

// Shuffle the cards
func (c *CardPile) Shuffle() {
	var i = uint64(len(c.MCards) - 1)
	for ; i > 0; i-- {
		j := rand_range(i + 1)
		c.MCards[i], c.MCards[j] = c.MCards[j], c.MCards[i]
	}
}

// Refresh the cardpile
func (c *CardPile) Refresh() {
	c.MCards = c.MCards[:0]
	c.MCards = append(c.MCards, c.mOriginalCards...)
}

File: CBlackJack/src/cardpile.c
#include "cardpile.h"

#include <stdint.h>
#include <stdio.h>
#include <time.h>

#include "card.h"
#include "vector.h"

char* ranks[] = {"A", "2", "3",  "4", "5", "6", "7",
                 "8", "9", "10", "J", "Q", "K"};
char* suits[] = {"Clubs", "Hearts", "Spades", "Diamonds"};

uint64_t state;

// From https://github.com/lemire/testingRNG
uint64_t wyrand() {
  state += UINT64_C(0xa0761d6478bd642f);
  __uint128_t t = (__uint128_t)state * (state ^ UINT64_C(0xe7037ed1a0b428db));
  return (t >> 64) ^ t;
}

// use nearly divisionless technique found here
// https://github.com/lemire/FastShuffleExperiments
uint64_t rand_range(uint64_t s) {
  uint64_t x = wyrand();
  __uint128_t m = (__uint128_t)x * (__uint128_t)s;
  uint64_t l = (uint64_t)m;
  if (l < s) {
    uint64_t t = -s % s;
    while (l < t) {
      x = wyrand();
      m = (__uint128_t)x * (__uint128_t)s;
      l = (uint64_t)m;
    }
  }
  return m >> 64;
}

CardPile* CardPile__new(int numdecks) {
  state = time(NULL);
  CardPile* cp = malloc(sizeof(CardPile));
  cp->m_original_cards = Vector__new(52 * numdecks);
  cp->m_cards = Vector__new(52 * numdecks);
  for (int i = 0; i < numdecks; i++) {
    for (int suit = 0; suit < 4; suit++) {
      for (int rank = 0; rank < 13; rank++) {
        Vector__push(cp->m_original_cards, Card__new(ranks[rank], suits[suit]));
      }
    }
  }
  CardPile__refresh(cp);
  return cp;
}

void CardPile__refresh(CardPile* self) {
  Vector__copy(self->m_original_cards, self->m_cards);
}

void CardPile__print(CardPile* self) {
  for (int i = 0; i < self->m_cards->size; i++) {
    printf("%s", ((Card*)self->m_cards->items[i])->m_rank);
  }
  printf("\n");
}

void CardPile__shuffle(CardPile* self) {
  for (int i = self->m_cards->size - 1; i > 0; i--) {
    int j = rand_range(i + 1);
    Card* temp = self->m_cards->items[i];
    self->m_cards->items[i] = self->m_cards->items[j];
    self->m_cards->items[j] = temp;
  }
}